### ISSUE TITLE
[all] Handle supervisor restart properly

### ISSUE THREAD

**Issue by mocchira** [leo-project/leofs#977]
As reported on https://github.com/leo-project/leofs/issues/966#issuecomment-358582625, the restarts of leo_mq_sup or leo_object_storage_sup might cause the supervisor process tree to be broken (part of worker processes failed to restart) the below is the crash.log when that happened.
```erlang
2018-01-10 20:23:50 =SUPERVISOR REPORT====
     Supervisor: {local,leo_storage_sup}
     Context:    child_terminated
     Reason:     shutdown
     Offender:   [{pid,<0.111.0>},{id,leo_object_storage_sup},{mfargs,{leo_object_storage_sup,start_link,[[{64,"/mnt/avs1/bodies"},{64,"/mnt/avs2/bodies"},{64,"/mnt/avs3/bodies"},{64,"/mnt/avs4/bodies"}],leo_storage_msg_collector]}},{restart_type,permanent},{shutdown,2000},{child_type,supervisor}]

2018-01-10 20:23:50 =ERROR REPORT====
{module,"leo_backend_db_eleveldb"},{function,"open/1"},{line,96},{body,{db_open,"IO error: lock /mnt/avs1/bodies/metadata/63/LOCK: already held by process"}}
2018-01-10 20:23:50 =CRASH REPORT====
  crasher:
    initial call: leo_backend_db_server:init/1
    pid: <0.17246.1505>
    registered_name: []
    exception exit: {{db_open,"IO error: lock /mnt/avs1/bodies/metadata/63/LOCK: already held by process"},[{gen_server,init_it,6,[{file,"gen_server.erl"},{line,344}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [leo_backend_db_sup,leo_object_storage_sup,leo_storage_sup,<0.46.0>]
    messages: []
    links: [<0.17234.1505>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 987
    stack_size: 27
    reductions: 371
  neighbours:
2018-01-10 20:23:50 =ERROR REPORT====
{module,"leo_backend_db_eleveldb"},{function,"open/1"},{line,96},{body,{db_open,"IO error: lock /mnt/avs2/bodies/metadata/10063/LOCK: already held by process"}}
2018-01-10 20:23:50 =CRASH REPORT====
  crasher:
    initial call: leo_backend_db_server:init/1
    pid: <0.17237.1505>
    registered_name: []
    exception exit: {{db_open,"IO error: lock /mnt/avs2/bodies/metadata/10063/LOCK: already held by process"},[{gen_server,init_it,6,[{file,"gen_server.erl"},{line,344}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [leo_backend_db_sup,leo_object_storage_sup,leo_storage_sup,<0.46.0>]
    messages: []
    links: [<0.17234.1505>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 987
    stack_size: 27
    reductions: 347
  neighbours:
```
Just in case, we have to test/check if all supervisor restarts work as expected.

**Comment by mocchira** [comment]
TODO:
- [ ] leo_mq_sup: needs FIX
- [ ] leo_backend_db_sup: needs FIX
- [ ] leo_object_storage_sup: needs FIX

**Comment by mocchira** [comment]
There are two possible fixes for the supervisor restart.
- Use supervisor2 https://github.com/rabbitmq/rabbitmq-common/blob/master/src/supervisor2.erl and specify the delay parameter then the supervisor restart itself will not happen anymore.
- Track all supervisor:start_child calls into the permanent ets table and restore all children based on that when the supervisor got restarted.

IMO, the former one would be promising as the latter costs so much to implement compared to the former and also it takes much time on the supervisor restart especially when the number of AVS files is large like @vstax use case. what do you think @yosukehara ?

if you agree with the former one, I will add the dependency to rabbitmq/rabbitmq-common and use supervisor2 with the delay instead.

**Comment by yosukehara** [comment]
@mocchira I've reviewed your opinion. I recommend it depends on [rabbitmq-common/master/src/supervisor2](https://github.com/rabbitmq/rabbitmq-common/blob/master/src/supervisor2.erl) because of keeping simple implementation.

**Comment by mocchira** [comment]
@yosukehara @windkit 
https://github.com/dLuna/chaos_monkey can be helpful to find this kind of problems in our code so I'd like to propose adding chaos_monkey to the dependencies and turn on (call chaos_monkey:on/2) while doing long running tests and so on.

**Comment by yosukehara** [comment]
@mocchira I've realized we use `chaos_monkey` to avoid similar error.

**Comment by mocchira** [comment]
It turned out that using stuffs in rabbitmq-common with rebar2 forces us to make a bit hack in rebar.config as it does't form a standard OTP directory structure just after downloaded (make to be invoked).

kind of the patch below (Combo using post_hooks and raw dependencies) might work
```diff
+{post_hooks, [{'get-deps', "cd deps/rabbitmq_common;make"}]}.
+
 {deps, [
-        {leo_backend_db, ".*", {git, "https://github.com/leo-project/leo_backend_db.git", {tag, "1.2.27"}}},
-        {leo_logger,     ".*", {git, "https://github.com/leo-project/leo_logger.git",     {tag, "1.3.5"}}}
+        {leo_backend_db,  ".*", {git, "https://github.com/leo-project/leo_backend_db.git", {tag, "1.2.27"}}},
+        {leo_logger,      ".*", {git, "https://github.com/leo-project/leo_logger.git",     {tag, "1.3.5"}}},
+        {rabbitmq_common, ".*", {git, "https://github.com/rabbitmq/rabbitmq-common.git",   {tag, "v3.7.3"}}, [raw]}
        ]}.
``` 
however I seem to have connection problem when fetching lager dependencies right now so postponed for now.

**Comment by mocchira** [comment]
> however I seem to have connection problem when fetching lager dependencies right now so postponed for now.

It turned out that the destination URL where a hex package is placed looks like below
```
https://s3.amazonaws.com/s3.hex.pm/tarballs/lager-3.5.1.tar
```
so they are placed on s3 however on my dev-box, s3.amazonaws.com has been resolved to 127.0.0.1 for testing whether various aws-sdk-clietnts works with LeoFS as expected so fetching hex dependencies silently failed.. Please be careful > devs
